### ISSUE TITLE
distill: train single-teacher on-policy distillation and report per-token and per-prompt

### ISSUE THREAD

**Issue by asuzukosi** [asuzukosi/ai-experiments#30]
The student generates, the teacher scores, and the student trains on its own best output. This should reduce the train/inference distribution mismatch that offline data suffers from.

Because this shares a prompt set with the offline baseline, report the comparison **per token and per prompt**. The on-policy premium often shrinks under prompt control, and **the two views disagreeing is itself the result worth publishing**.

Also produces the seed checkpoint the self-distillation arm needs, which is why it comes first.

### Sketch

```python
# src/distill/on_policy.py
def on_policy_round(student, teacher, prompts, n=8):
    samples = student_llm.generate([p.text for p in prompts],
                                   SamplingParams(n=n, temperature=1.0, max_tokens=1024))
    best = []
    for p, s in zip(prompts, samples):
        cands = [c.text for c in s.outputs]
        scores = teacher_score(teacher, p.text, cands)      # 1-10 rating, temperature 0
        best.append({"prompt": p.text, "completion": cands[argmax(scores)]})
    return best

TEACHER_SCORE_PROMPT = """Rate the response to the question on a scale of 1-10.
Reply with only the number.

Question: {question}
Response: {response}
Rating:"""

data = on_policy_round(student, teacher, prompts)   # shared tokenizer required
SFTTrainer(model=student, train_dataset=data, args=cfg).train()

# report BOTH views
log({"tokens_trained": total_tokens(data), "prompts_used": len(data)})
```

### TODO
- [ ] Student samples N=8 per prompt at temperature 1.0 over the same 50K prompts
- [ ] Teacher scores each sample (1-10, temperature 0); SFT on the argmax per prompt
- [ ] **Report per token and per prompt** — the two comparisons can disagree
- [ ] Teacher and student must share a tokenizer (both Qwen2.5)
- [ ] Record tokens trained and prompts used for both control conditions
- [ ] Save the resulting checkpoint — the self-distillation arm seeds from it

### References
- On-policy distillation of language models (GKD): https://arxiv.org/abs/2306.13649
- TRL SFTTrainer: https://huggingface.co/docs/trl/main/en/sft_trainer
- Rejection sampling / best-of-n fine-tuning, Llama 2: https://arxiv.org/abs/2307.09288

### Compute

1x A100 80GB (teacher) + 1x RTX 4090 (train). ~5 hours.

---

### Dependencies

- **Blocked by:** #28
- **Blocks:** #31, #32, #33

Self-contained within this project — no cross-project dependencies.
